### ISSUE TITLE
WooCommerce breadcrumb ignores primary category

### ISSUE THREAD

**Issue by sukafia** [woothemes/woocommerce#31411]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When a product is assigned to multiple categories/subcategories, WooCommerce doesn't use the primary category for breadcrumbs. Instead, it uses the 2nd category.

Forum threads: 

- https://wordpress.org/support/topic/breadcrumbs-for-products-in-multiple-categories/
- https://wordpress.org/support/topic/wrong-breadcrumbs-at-multiple-category-product/

While the issue is attributed to how WordPress handles breadcrumbs, I've shared why I think otherwise in the steps to replicate the issue.

### Expected behavior

When you're visiting a product page directly (not from a certain category), the primary category should be used for breadcrumbs. 

### Actual behavior

When you visit a product directly, the 2nd category/sub-category is used for breadcrumbs. 

### Steps to reproduce

1. Create a parent & child category (Family > Mother)
2. Create another parent & child category (Friends > Sunday)
3. Create a product & assign it to the "Mother" sub-category
4. Also assign the product to the "Sunday" sub-category
5. Set "Mother" as the product's primary category
6. Publish the product & visit the product page through the product's direct link (not through the category)
6. Breadcrumbs ignores the primary category & uses the non-primary category "Friends > Damianne" instead of "Family > Mother".

It's been argued that this is due to how WordPress handles breadcrumbs but I don't think so. You can confirm this by repeating the above steps for a WordPress post and you'll see posts recognize the primary category and use it for breadcrumbs.

### WordPress Environment

`
### WordPress Environment ###

WC Version: 5.9.0
REST API Version: ✔ 5.9.0
WC Blocks Version: ✔ 6.1.0
Action Scheduler Version: ✔ 3.3.0
WC Admin Version: ✔ 2.9.1
Log Directory Writable: ✔
WP Version: 5.8.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 7.4.26
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 7.78.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.3.24-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.9.0
WC Database Prefix: wp_
Total Database Size: 52.01MB

Tested on Storefront, Twenty Twenty & Twenty Twenty-One.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by swatipawarGS** [comment]
Hi @sukafia ,

Thank you for taking the time to report this bug, we really appreciate your help. We are able to reproduce the issue on our end using WooCommerce 5.9.0 version. 

Please find below screenshots  for reference :

![#31411(cat)](https://user-images.githubusercontent.com/94531721/145944601-4f9ff661-5901-4a91-98dc-f3cd72c33c7a.jpg)


![#31411(productpage)](https://user-images.githubusercontent.com/94531721/145944607-2d7106ce-b886-4d89-af85-cdca7c7df74c.jpg)

Priority of this issue will be assigned shortly after we’d go through a round of prioritization. Please note it may take a couple of days.

**Comment by rodelgc** [comment]
Hello @sukafia,

I just want to clarify some of the steps to make sure that I and Swati were able to reproduce it correctly:

> 5. Set "Mother" as the product's primary category

I suppose this should be _"Family"_?

> 7. Breadcrumbs ignores the primary category & uses the non-primary category "Friends > Damianne" instead of "Family > Mother".

I assume you meant  _"Friends > Sunday"_?

If so, I tried reproducing the issue using these assumptions and came up with the following category setup for my product:

<img width="304" alt="Screen Shot 2021-12-17 at 8 03 04 PM" src="https://user-images.githubusercontent.com/4509348/146542147-26360eaf-769b-47b7-9780-86dc5717d0fa.png">

☝️ This is slightly different from Swati's screenshot. However, I was able to see the same erroneous breadcrumb when I:

1. Navigated directly to the product page using the permalink:
 
<img width="680" alt="Screen Shot 2021-12-17 at 8 05 45 PM" src="https://user-images.githubusercontent.com/4509348/146543132-22a46488-fab9-4a14-9dd3-8f4c8bd54f61.png">

<img width="738" alt="Screen Shot 2021-12-17 at 8 07 26 PM" src="https://user-images.githubusercontent.com/4509348/146543145-ff0b9734-0ae9-480d-831d-37f4939330d0.png">

2. Navigated to the product page from the `/family/mother` category page:

<img width="704" alt="Screen Shot 2021-12-17 at 8 10 26 PM" src="https://user-images.githubusercontent.com/4509348/146543204-0b805909-769f-4743-a123-0eee1a1d5b9e.png">

I'm labeling this now with `type: bug` since we're able to encounter the bug. But I'd still wait for your confirmation if we got the reproduction steps correctly. Thanks!

**Comment by sukafia** [comment]
Both steps are correct. The main issue is that WC is ignoring the product's primary category (both for category & sub-category). In my steps, I used the sub-category but the issue exists for both.

If I set a product as the primary category & visit the product using the [direct URL](https://user-images.githubusercontent.com/4509348/146543132-22a46488-fab9-4a14-9dd3-8f4c8bd54f61.png) it should use the primary category in the breadcrumbs & not the secondary category.

**Comment by ghost** [comment]
Hi, may i know when will this issue be fixed? It's kinda hurting my site SEO and site navigation
![breadcrumb](https://user-images.githubusercontent.com/96566870/147204654-44a30061-1819-46d5-9451-1e763474b3a1.JPG)
![site breadcrumb](https://user-images.githubusercontent.com/96566870/147204815-25452648-b630-4b09-9239-1b8adec2abd6.JPG)

They are showing the category that aren't set as primary category which is supposed to be **Hydrators / Toners /** Olivarrier Dual Moist Toning Lotion

Thanks~!

**Comment by aheckler** [comment]
4630999-zen

**Comment by ghost** [comment]
Why hasn't the bug been fixed? I switched to a new theme in hope that I could finally have more categories but :\

**Comment by Jas657** [comment]
This issue is still there, when can this expect to be fixed?

**Comment by CW2000** [comment]
We'd like to see this fixed too.

**Comment by beezwings** [comment]
Yes! Any news on when this might be fixed?

**Comment by JakeDevLab** [comment]
Hello from 2023 (more than 2 years after the bug was reported).
Will there be a hero who fixed the bug one day? :)

**Comment by ricardoaraujo330** [comment]
Please fix it! Main category is being ignored. Breadcrumbs should show the first category based on the category list order!

**Comment by shorshe** [comment]
This seems to be a long-standing Bug with no easy workarounds. This hurts SEO-Scores. I'd really love to see this fixed. Is there a way to bump up the priority?

**Comment by xristosstamos** [comment]
I have encountered the same problem too.
If anyone has found a solution I would love to try it out

**Comment by johnlomat** [comment]
This bug is still not fixed? Any workaround?

**Comment by CameronWhyte** [comment]
We're still waiting on a fix for this issue too. The path of the product breadcrumb just seems to be completely random, would be ideal if it used the path to the primary category.

**Comment by ghost** [comment]
Still there in November 2023

**Comment by theo-gk** [comment]
Still an issue after all these years. Could you please implement a fix? This is serious. Thanks

**Comment by sefaraah121** [comment]
I'm having the same problem. Thought it was my theme or the Yoast plugin, but found a workaround with the code below (added to functions.php). The breadcrumb will pick up the primary category and related subcategories, **HOWEVER**, the URL won't reflect this, so still a major issue

**EDIT: Tried to get the entire code on the plain text section, but for some reason this isn't working. Please check carefully if copying.**


`add_filter('woocommerce_get_breadcrumb', 'custom_woocommerce_breadcrumbs', 10, 2);
function custom_woocommerce_breadcrumbs($crumbs, $breadcrumb) {
    if (is_product()) {
        global $post;
        $primary_category_id = yoast_get_primary_term_id('product_cat', $post->ID);

        if ($primary_category_id) {
            // Fetch the primary category details
            $primary_category = get_term($primary_category_id, 'product_cat');

            if ($primary_category && !is_wp_error($primary_category)) {
                $updated_crumbs = array($crumbs[0]);

                if ($primary_category->parent != 0) {
                    $parent_category = get_term($primary_category->parent, 'product_cat');
                    if ($parent_category && !is_wp_error($parent_category)) {
                        $updated_crumbs[] = array($parent_category->name, get_term_link($parent_category));
                    }
                }

                $updated_crumbs[] = array($primary_category->name, get_term_link($primary_category));

                $updated_crumbs[] = end($crumbs);

                return $updated_crumbs;
            }
        }
    }

    return $crumbs;
}
`

**Comment by CameronWhyte** [comment]
@sefaraah121 your code above is working perfectly on our site, many thanks for posting!

We also use the Permalink Manager Pro plugin which allows us to set our product URL's, so that negates the URL issue you mention.

**Comment by sefaraah121** [comment]
Perfect! Glad it worked for you too :)

**Comment by Jameswilliam1122** [comment]
> Hi @sukafia ,
> 
> Thank you for taking the time to report this bug, we really appreciate your help. We are able to reproduce the issue on our end using WooCommerce 5.9.0 version.
> 
> Please find below screenshots for reference :
> 
> ![#31411(cat)](https://user-images.githubusercontent.com/94531721/145944601-4f9ff661-5901-4a91-98dc-f3cd72c33c7a.jpg)
> 
> ![#31411(productpage)](https://user-images.githubusercontent.com/94531721/145944607-2d7106ce-b886-4d89-af85-cdca7c7df74c.jpg)
> 
> Priority of this issue will be assigned shortly after we’d go through a round of prioritization. Please note it may take a couple of days.

Its 2024 is not fixed yet.

**Comment by BTB62** [comment]
I'm having the same problem.  The path of the product breadcrumb just seems to be completely random, would be ideal if it used the path to the primary category.

Any update on this issue. Using Beaver Builder and Rank Math ?

**Comment by netconnectsolutions** [comment]
I'm having this issue too

**Comment by gaerd** [comment]
pls fix this

**Comment by ricardoaraujo330** [comment]
When will it be fixed?

**Comment by md-sherazul-islam** [comment]
@sefaraah121 Thank you for the fix. It worked for me. 

However,  I've modified it a little bit to show all the parent categories it has. Check this out - 

add_filter('woocommerce_get_breadcrumb', 'sp_custom_woocommerce_breadcrumbs', 10, 2);
function sp_custom_woocommerce_breadcrumbs($crumbs, $breadcrumb) {
    if (is_product()) {
        global $post;
        $primary_category_id = yoast_get_primary_term_id('product_cat', $post->ID);


        if ($primary_category_id) {
            // Fetch the primary category details
            $primary_category = get_term($primary_category_id, 'product_cat');


            if ($primary_category && !is_wp_error($primary_category)) {
                $updated_crumbs = array($crumbs[0]);


                // Traverse up the category hierarchy and add each parent category to the breadcrumbs
                $parent_category = $primary_category;
                while ($parent_category->parent != 0) {
                    $parent_category = get_term($parent_category->parent, 'product_cat');
                    if ($parent_category && !is_wp_error($parent_category)) {
                        array_splice($updated_crumbs, 1, 0, array(array($parent_category->name, get_term_link($parent_category))));
                    }
                }


                $updated_crumbs[] = array($primary_category->name, get_term_link($primary_category));
                $updated_crumbs[] = end($crumbs);


                return $updated_crumbs;
            }
        }
    }


    return $crumbs;
}

**Comment by epnb** [comment]
Hello, 
The fix we are using in the filter, which might help others, involves sorting by URL length, from shortest (ancestor) to longest (deepest category)

`        usort($crumbs, function($a, $b) {
            $aLength = strlen($a[1])>0?strlen($a[1]):99999;
            $bLength = strlen($b[1])>0?strlen($b[1]):99999;
            return $aLength - $bLength;
        });`

**Comment by mdperez86** [comment]
WooCommerce does not currently offer a built-in feature to select a primary category for products. However, it provides filters that can be used to customize this behavior, such as [woocommerce_breadcrumb_product_terms_args](https://github.com/woocommerce/woocommerce/blob/ef015f740f89c35cebf34bc0c39a971be6723b68/plugins/woocommerce/includes/class-wc-breadcrumb.php#L154) to order categories and [woocommerce_breadcrumb_main_term](https://github.com/woocommerce/woocommerce/blob/ef015f740f89c35cebf34bc0c39a971be6723b68/plugins/woocommerce/includes/class-wc-breadcrumb.php#L163) to select the main category. The default WooCommerce logic sorts categories by parent ID and selects the latest one, which explains why breadcrumbs always display a child category of the most recently created parent category.

Given that WooCommerce only provides basic category sorting logic in this context, I see this as a feature request rather than a bug. If a specific plugin, such as Yoast SEO, allows users to select a primary category, it is responsible for ensuring that feature integrates properly with WooCommerce, including in breadcrumbs.

Let me know if you need further adjustments!

**Comment by secure11** [comment]
Can anyone repost @sefaraah121 php code that was reported to be working to use the yoast primary category for the product page breadcrumb? At least for me, there are some threads in this topic that won't display and that is one of them so I don't see the php code. Thanks

**Comment by mdperez86** [comment]
As referenced in my [previous comment](https://github.com/woocommerce/woocommerce/issues/31411#issuecomment-2371750759), this is not a WooCommerce bug, as the concept of a “primary category” is not part of WooCommerce’s core functionality. Instead, it is a feature provided by the Yoast SEO plugin.

For those looking to implement a solution that reflects the behavior described here, you can achieve this by utilizing WooCommerce’s breadcrumb-related hooks. Below is an example code snippet that leverages WooCommerce hooks to display the Yoast primary category in the breadcrumbs.

Example Code Snippet:

This code uses the [woocommerce_breadcrumb_main_term](https://github.com/woocommerce/woocommerce/blob/ef015f740f89c35cebf34bc0c39a971be6723b68/plugins/woocommerce/includes/class-wc-breadcrumb.php#L163) filter to modify the primary category displayed in the WooCommerce breadcrumb based on Yoast’s settings. Please note, the Yoast SEO plugin must be installed and active for this functionality to work.

```php
add_filter( 'woocommerce_breadcrumb_main_term', 'get_yoast_primary_category_as_breadcrumb_main_term', 10, 2 );

function get_yoast_primary_category_as_breadcrumb_main_term( $main_term, $terms ) {
	// Retrieve the primary category ID set by Yoast for the current product.
	$primary_category_id = yoast_get_primary_term_id( 'product_cat', get_the_ID() );
	
	// Loop through the assigned terms and match the primary category.
	foreach ( $terms as $term ) {
		if ( $primary_category_id === $term->term_id ) {
			return $term;
		}
		// For cases where the primary category is a parent term.
		if ( $primary_category_id === $term->parent ) {
			// Return the first matching child term found in the terms array.
			return $term;
		}
	}
	
	return $main_term;
}
```

To implement this on your site, simply add the above code using a code snippet plugin, which allows custom code to run safely without modifying theme or plugin files directly. This approach ensures compatibility and maintainability for future updates.

**Comment by j-sanch** [comment]
This worked for me. Click the "QuickEdit" button and remove the category you want to be the primary category then save. Then click "Quick Edit" again.  Re-Check the category and save. It made the correct category appear in the breadcrumb for me.

**Comment by mdperez86** [comment]
After thorough internal discussions, we have decided to mark this issue as `won't fix`. The primary concern raised here relates to the main category functionality introduced by the Yoast SEO plugin, rather than a WooCommerce core feature. Since this behavior is not controlled by WooCommerce, it falls outside the scope of our core development efforts.

However, we understand that this may still be an important requirement for some users. As a workaround, you can refer to the code snippet shared in [this comment](https://github.com/woocommerce/woocommerce/issues/31411#issuecomment-2468525696), which may help achieve the desired functionality.

We appreciate your feedback and contributions to WooCommerce, and we encourage further discussions or alternative solutions from the community.

pdnLyh-7pj-p2

**Comment by ricardoaraujo330** [comment]
First of all, I appreciate finally receiving an official response after 4 years! I’m sure other users will feel the same.

However, I strongly disagree with the decision to mark this as "won't fix". The original report and multiple user confirmations refer specifically to WooCommerce’s breadcrumbs, not Yoast SEO.

WooCommerce itself provides an option to set a primary category in the product category selection UI. If this functionality exists within WooCommerce, then logically, it should be reflected in the frontend breadcrumbs, without requiring additional code or reliance on third-party plugins.

This seems like a minor and easy fix that would significantly improve WooCommerce’s default behavior. Marking it as out of scope does not align with the fact that WooCommerce already allows defining a primary category.

We believe this decision deserves to be reconsidered, and we urge the team to address it properly instead of requiring users to implement workarounds.

**Comment by JakeDevLab** [comment]
@mdperez86 please could you see the comment written by @ricardoaraujo330 and advise?

I wholeheartedly support the idea that since WooCommerce offers an option to set a primary category in the product category selection UI, it should be reflected in the frontend breadcrumbs. This functionality should be integrated seamlessly without the need for additional code or external third-party plugins.
